airbyte-integrations/connectors/source-chargebee/unit_tests/integration/test_coupon.py

```diff
@@ -0,0 +1,208 @@
+            ],
+        )
+        output = self._read(_config().with_start_date(self._start_date))
+        assert len(output.records) == 1
+
+    @HttpMocker()
+    def test_given_http_status_500_once_before_200_when_read_then_retry_and_return_records(self, http_mocker: HttpMocker) -> None:
+        # Tests retry with 500 status
+        http_mocker.get(
+            _a_request().with_any_query_params().build(),
+            [a_response_with_status_and_header(500, {"Retry-After": "0.01"}), _a_response().with_record(_a_record()).build()],
+        )
+        output = self._read(_config().with_start_date(self._start_date))
+        assert len(output.records) == 1
+
+    @HttpMocker()
+    def test_given_http_status_500_on_availability_when_read_then_raise_system_error(self, http_mocker: HttpMocker) -> None:
+        # Tests 500 status error handling
+        http_mocker.get(
+            _a_request().with_any_query_params().build(),
+            a_response_with_status_and_header(500, {"Retry-After": "0.01"}),
+        )
+        output = self._read(_config(), expecting_exception=True)
+        assert output.errors[-1].trace.error.failure_type == FailureType.system_error
+
+@freezegun.freeze_time(_NOW.isoformat())
+class IncrementalTest(TestCase):
+
+    def setUp(self) -> None:
+        self._now = _NOW
+        self._now_in_seconds = int(self._now.timestamp())
+        self._start_date = _NOW - timedelta(days=60)
+        self._start_date_in_seconds = int(self._start_date.timestamp())
+
+    @staticmethod
+    def _read(config: ConfigBuilder, state: Dict[str, Any], expecting_exception: bool = False) -> EntrypointOutput:
+        return _read(config, SyncMode.incremental, state, expecting_exception=expecting_exception)
+
+    @HttpMocker()
+    def test_given_no_initial_state_when_read_then_return_state_based_on_most_recently_read_slice(self, http_mocker: HttpMocker) -> None:
+        # Tests setting state when no initial state is provided
+        cursor_value = self._start_date_in_seconds + 1
+        http_mocker.get(
+            _a_request().with_any_query_params().build(),
+            _a_response().with_record(_a_record().with_cursor(cursor_value)).build()
+        )
+        output = self._read(_config().with_start_date(self._start_date - timedelta(hours=8)), _NO_STATE)
+        assert output.most_recent_state == { _STREAM_NAME: {_CURSOR_FIELD: str(self._now_in_seconds) }}
+
+    @HttpMocker()
+    def test_given_initial_state_use_state_for_query_params(self, http_mocker: HttpMocker) -> None:
+        # Tests updating query param with state
+        state_cursor_value = int((self._now - timedelta(days=5)).timestamp())
+        state =  StateBuilder().with_stream_state(_STREAM_NAME, {_CURSOR_FIELD: state_cursor_value}).build()
+        http_mocker.get(
+            _a_request().with_sort_by_asc("created_at").with_include_deleted(True).with_updated_at_btw([state_cursor_value, self._now_in_seconds]).build(),
```

Comment:
Why do we sort by `created_at`? I don't see any value in this and this could lead to a false sense of "we can enable checkpoint interval because we sort". This would not be true as demonstrated by the following sequence of events:

T0: X is created
T1: Y is created
T2: X is updated
T3: Start a sync, consume X, emit state T2 based on checkpoint interval, fail between emitting X and Y

In this case, Y would not be synced as the state T2 is greater than T1, the updated_at value for Y

Reply:
Yes I agree with this. I have removed `created_at` sorting for this stream.